src/dataflow/src/source/mod.rs

```diff
@@ -1784,9 +1776,36 @@ impl SourceReaderPersistence {
                 *diff == 1
             })
             .map(|(binding, _diff)| binding)
-            .collect();
+            .collect::<Vec<_>>();
 
-        Ok((starting_offsets, bindings))
+        // For retractions, we just need the "consolidated" diff value for each binding/timestamp.
+        let retractions = retractions
+            .drain()
+            .filter(|(_binding, diff)| diff != &0)
```

Comment:
should we do the same check that the diff == 1 here?

Reply:
I'll add it, but it has to be `diff.abs()` because it can be negative. :+1: